File: 2023/d06/day.rb
file = "input.txt"
times, distances = []

times, distances = File.readlines(file).map do |l|
  l.strip.split(":").last.split(" ").map(&:to_i)
end

t2, d2 = File.readlines(file).map do |l|
  l.strip.split(":").last.gsub(/\s+/, "").to_i
end

times << t2
distances << d2

p1 = times.map.with_index do |time, i|
  (1..time).map do |t|
    t * (time - t)
  end.select{|n| n > distances[i]}.count
end
p2 = p1.pop
puts p1.inject(:*)
puts p2

File: 2023/d06/day.exs
lines =
  File.read!("input.txt")
  |> String.split("\n", trim: true)

[times, distances] =
  lines
  |> Enum.map(fn line ->
    line
    |> String.split(":")
    |> List.last()
    |> String.split(" ", trim: true)
    |> Enum.map(&String.to_integer/1)
  end)

[t2, d2] =
  lines
  |> Enum.map(fn line ->
    line
    |> String.split(":")
    |> List.last()
    |> String.replace(~r/\s+/, "")
    |> String.to_integer()
  end)

times = [t2 | times]
distances = [d2 | distances]

[p2 | nb_of_ways_to_beat_record] =
  times
  |> Enum.with_index()
  |> Enum.map(fn {time, i} ->
    1..time
    |> Enum.reduce(0, fn t, acc ->
      if t * (time - t) > Enum.at(distances, i),
        do: acc + 1,
        else: acc
    end)
  end)

nb_of_ways_to_beat_record
|> Enum.reduce(1, fn n, acc -> n * acc end)
|> IO.puts()

IO.puts(p2)
